internal/planner/operation.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package planner
+
+import (
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/internal/core"
+	"github.com/sourcenetwork/defradb/internal/planner/mapper"
+)
+
+const operationNodeKind string = "operationNode"
+
+// operationNode is the top level node for operations with
+// one or more child selections, such as queries or mutations.
+type operationNode struct {
+	documentIterator
+	docMapper
+
+	children []planNode
+	isDone   bool
+}
+
+func (n *operationNode) Spans(spans core.Spans) {
+	for _, child := range n.children {
+		child.Spans(spans)
+	}
+}
+
+func (n *operationNode) Kind() string {
+	return operationNodeKind
+}
+
+func (n *operationNode) Init() error {
+	n.isDone = false
+	for _, child := range n.children {
+		err := child.Init()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *operationNode) Start() error {
+	for _, child := range n.children {
+		err := child.Start()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *operationNode) Close() error {
+	for _, child := range n.children {
+		err := child.Close()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *operationNode) Source() planNode {
+	return nil
+}
+
+func (p *operationNode) Children() []planNode {
+	return p.children
+}
+
+func (n *operationNode) Explain(explainType request.ExplainType) (map[string]any, error) {
+	switch explainType {
+	case request.SimpleExplain:
+		return map[string]any{}, nil
+
+	case request.ExecuteExplain:
+		return map[string]any{}, nil
+
+	default:
+		return nil, ErrUnknownExplainRequestType
+	}
+}
+
+func (n *operationNode) Next() (bool, error) {
+	if n.isDone {
+		return false, nil
+	}
+
+	n.currentValue = n.documentMapping.NewDoc()
+	for i, child := range n.children {
+		switch child.(type) {
+		case *topLevelNode:
+			hasChild, err := child.Next()
+			if err != nil {
+				return false, err
+			}
+			if !hasChild {
+				return false, ErrMissingChildValue
+			}
+			n.currentValue = child.Value()
+
+		default:
+			var docs []core.Doc
+			for {
+				hasChild, err := child.Next()
+				if err != nil {
+					return false, err
+				}
+				if !hasChild {
+					break
+				}
+				docs = append(docs, child.Value())
+			}
+			n.currentValue.Fields[i] = docs
+		}
+	}
+
+	n.isDone = true
```

Comment:
question: do you really need this flag? It's being set only in `Init()`. `Next()` is not supposed to be called anyway before `Init()` and it's not multi-threaded.

Reply:
I tried removing it, but the return value was always incorrect. I believe we need it so that the operation is able to run at least once.

Reply:
nitpick/todo: The planNode interface and it's consuption dictates that `Init` can be called multiple times, and that it's state should be reset by the first call (before doing it's thing again).  I can't think why this node would have an instance Init'ed multiple times, but it is kind of incorrect atm, and just last week we had a production issue caused by this.

Reply:
Should be fixed now. I'm assuming you were referencing the `currentValue` not being reset.